crates/bevy_audio/src/audio_output.rs

```diff
@@ -22,157 +22,213 @@ use crate::AudioSink;
+                                    // PERF: insert as bundle to reduce archetype moves
+                                    .insert((AudioSink { sink: Some(sink) }, PlaybackRemoveMarker));
+                            }
+                        };
                     }
-                } else if let Some(sink) = self.play_source(audio_source, config.settings.repeat) {
-                    sink.set_speed(config.settings.speed);
-
-                    match config.settings.volume {
-                        Volume::Relative(vol) => sink.set_volume(vol.0 * global_volume.volume.0),
-                        Volume::Absolute(vol) => sink.set_volume(vol.0),
+                    Err(err) => {
+                        warn!("Error playing sound: {err:?}");
                     }
-
-                    // don't keep the strong handle. there is no way to return it to the user here as it is async
-                    let _ = sinks.set(config.sink_handle, AudioSink { sink: Some(sink) });
                 }
-            } else {
-                // audio source hasn't loaded yet. add it back to the queue
-                queue.push_back(config);
             }
-            i += 1;
         }
     }
 }
 
-/// Plays audio currently queued in the [`Audio`] resource through the [`AudioOutput`] resource
-pub fn play_queued_audio_system<Source: Asset + Decodable>(
-    audio_output: Res<AudioOutput<Source>>,
-    audio_sources: Option<Res<Assets<Source>>>,
-    global_volume: Res<GlobalVolume>,
-    mut audio: ResMut<Audio<Source>>,
-    mut sinks: ResMut<Assets<AudioSink>>,
-    mut spatial_sinks: ResMut<Assets<SpatialAudioSink>>,
-) where
-    f32: rodio::cpal::FromSample<Source::DecoderItem>,
-{
-    if let Some(audio_sources) = audio_sources {
-        audio_output.try_play_queued(
-            &*audio_sources,
-            &mut *audio,
-            &mut sinks,
-            &mut spatial_sinks,
-            &global_volume,
-        );
-    };
+pub(crate) fn cleanup_finished_audio(
+    mut commands: Commands,
+    query_nonspatial_despawn: Query<(Entity, &AudioSink), With<PlaybackDespawnMarker>>,
+    query_spatial_despawn: Query<(Entity, &SpatialAudioSink), With<PlaybackDespawnMarker>>,
+    query_nonspatial_remove: Query<(Entity, &AudioSink), With<PlaybackRemoveMarker>>,
+    query_spatial_remove: Query<(Entity, &SpatialAudioSink), With<PlaybackRemoveMarker>>,
+) {
+    for (entity, sink) in &query_nonspatial_despawn {
+        if sink.sink.as_ref().unwrap().empty() {
+            commands.entity(entity).despawn();
+        }
+    }
+    for (entity, sink) in &query_spatial_despawn {
+        if sink.sink.as_ref().unwrap().empty() {
+            commands.entity(entity).despawn();
+        }
+    }
+    for (entity, sink) in &query_nonspatial_remove {
+        if sink.sink.as_ref().unwrap().empty() {
+            commands
+                .entity(entity)
+                .remove::<(AudioBundle, AudioSink, PlaybackRemoveMarker)>();
```

Comment:
AudioBundle / SpatialAudioBundle are currently defaulting to `AudioSource` here, which means that AudioSink and SpatialAudioSink entities with custom Source / `Decodable` audio types will have the wrong bundle removed.

To resolve this I think:
1. This system needs to be generic on Source
2. The queries need to filter on `With<Handle<Source>>`